lib/ace/ext/papeeria/latex_behaviour.coffee

```diff
@@ -0,0 +1,187 @@
+define((require, exports, module) ->
+  Behaviour = require("ace/mode/behaviour").Behaviour
+
+  isCommentToken = (token) -> token? and /comment/.test(token.type)
+
+  isEscaped = (line, column) -> line[column - 1] == '\\' and line[column - 2] != '\\'
+
+  isInEquation = (session, row, column) ->
+    token = session.getTokenAt(row, column)
+    nextToken = session.getTokenAt(row, column + 1)
+    state = session.getState(row)
+    pState = session.getState(row - 1)
+    lastState = if typeof(state) == "string" then state else state[state.length - 1]
+    lastPrevState = if typeof(pState) == "string" then pState else pState[pState.length - 1]
+    return (
+      (not token? and /math/.test(lastPrevState)) or
+      (token? and /equation/.test(token.type)) or
+      (nextToken? and /equation/.test(nextToken.type)) or
+      (not nextToken? and /math/.test(lastState))
+    )
+
+  dollarsInsertionAction = (state, action, editor, session, text) ->
+    if text == '$' and not editor.inMultiSelectMode
+      autoInsert = { text: "$$", selection: [1, 1] }
+      skip = { text: "", selection: [1, 1] }
+      doNothing = null
+
+      { row, column } = editor.getCursorPosition()
+      line = session.getLine(row)
+
+      selection = editor.getSelectionRange()
+      selected = session.getTextRange(selection)
+      if selected != ""
+        if editor.getWrapBehavioursEnabled()
+          return getWrapped(selection, selected, text, text)
+        else
+          return doNothing
+
+      token = session.getTokenAt(row, column)
+      nextToken = session.getTokenAt(row, column + 1)
+
+      if isCommentToken(token) and column != 0 or isEscaped(line, column)
+        return doNothing
+
+      prevChar = line[column - 1] or ''
+      nextChar = line[column] or ''
+
+      if isInEquation(session, row, column)
+        return if nextChar == '$' then skip else doNothing
+
+      shouldSkip = (nextChar == '$' and (prevChar != '$' or /rparen/.test(nextToken.type)))
+      return if shouldSkip then skip else autoInsert
+
+  dollarsDeletionAction = (state, action, editor, session, range) ->
+    selected = session.doc.getTextRange(range)
```

Comment:
@yanenkoa I don't know how that would affect logic of your code but aren't "multi line selection" and "multi select mode" different things? I'd expect "multi select mode" to mean "multiple cursors"